iree/tools/utils/image_util.c

```diff
@@ -0,0 +1,192 @@
+      stbi_load(filename, img_dims, &(img_dims[1]), &(img_dims[2]), req_ch);
+  if (*pixel_data == NULL) {
+    return iree_make_status(IREE_STATUS_NOT_FOUND, "can't load image %s",
+                            filename);
+  }
+  *buffer_length =
+      img_dims[0] * img_dims[1] * (img_dims[2] > 3 ? 3 : img_dims[2]);
+  return iree_ok_status();
+}
+
+iree_status_t iree_tools_utils_buffer_view_from_image(
+    const iree_hal_dim_t* shape, iree_host_size_t shape_rank,
+    iree_hal_element_type_t element_type, iree_hal_allocator_t* allocator,
+    const char* filename, iree_hal_buffer_view_t** buffer_view) {
+  *buffer_view = NULL;
+  iree_status_t result;
+  uint8_t* pixel_data = NULL;
+  iree_host_size_t buffer_length;
+  result = iree_tools_utils_load_pixel_data(
+      shape, shape_rank, element_type, filename, &pixel_data, &buffer_length);
+  if (!iree_status_is_ok(result)) {
+    return result;
+  }
+
+  iree_hal_buffer_t* buffer = NULL;
+  iree_host_size_t element_byte = iree_hal_element_byte_count(element_type);
+  switch (element_type) {
+    // SINT_8 and UINT_8 perform direct buffer wrap.
+    case IREE_HAL_ELEMENT_TYPE_UINT_8:
+    case IREE_HAL_ELEMENT_TYPE_SINT_8: {
+      result = iree_hal_buffer_view_wrap_or_clone_heap_buffer(
+          allocator, shape, shape_rank, element_type,
+          IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+          IREE_HAL_MEMORY_ACCESS_READ, IREE_HAL_BUFFER_USAGE_ALL,
+          iree_make_byte_span((void*)pixel_data, element_byte * buffer_length),
+          iree_allocator_null(), buffer_view);
+      break;
+    }
+    case IREE_HAL_ELEMENT_TYPE_FLOAT_32: {
+      result = iree_hal_allocator_allocate_buffer(
+          allocator,
+          IREE_HAL_MEMORY_TYPE_HOST_LOCAL | IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+          IREE_HAL_BUFFER_USAGE_ALL, element_byte * buffer_length, &buffer);
+      if (!iree_status_is_ok(result)) {
+        free(pixel_data);
+        return result;
+      }
+      // Need to normalize to the expected input range. Default to [0, 1].
+      iree_hal_buffer_mapping_t mapped_memory;
+      result = iree_hal_buffer_map_range(
+          buffer, IREE_HAL_MEMORY_ACCESS_DISCARD_WRITE, 0,
+          element_byte * buffer_length, &mapped_memory);
+      if (!iree_status_is_ok(result)) {
+        iree_hal_buffer_release(buffer);
+        free(pixel_data);
+        return result;
+      }
+      result = iree_tools_utils_pixel_rescaled_to_buffer(
+          pixel_data, buffer_length, (float*)mapped_memory.contents.data);
+      if (!iree_status_is_ok(result)) {
+        iree_hal_buffer_release(buffer);
+        free(pixel_data);
+        return result;
+      }
+      iree_hal_buffer_unmap_range(&mapped_memory);
+      result = iree_hal_buffer_view_create(buffer, shape, shape_rank,
+                                           element_type, buffer_view);
+      iree_hal_buffer_release(buffer);
+      break;
+    }
+    default: {
+      free(pixel_data);
```

Comment:
Could use `stbi_image_free` in this file instead of `free`, right?

Reply:
From stb_image.h
```
#ifndef STBI_MALLOC
#define STBI_MALLOC(sz)           malloc(sz)
#define STBI_REALLOC(p,newsz)     realloc(p,newsz)
#define STBI_FREE(p)              free(p)
#endif


STBIDEF void stbi_image_free(void *retval_from_stbi_load)
{
   STBI_FREE(retval_from_stbi_load);
}
```
Updated so we have the consistency.